### ISSUE TITLE
AssetsDefinition.from_graph is missing name argument

### ISSUE THREAD

**Issue by simonvanderveldt** [dagster-io/dagster#12352]
### What's the use case?

`AssetsDefinition.from_graph()` doesn't have a name argument. This means it's impossible to define the name of the Assets that's created from this.

This is problematic because graphs need to have a unique name (per repository as far as I'm aware, see https://github.com/dagster-io/dagster/issues/8076) and they don't have support for something like multi-component Asset/Graph keys (`key_prefix`) to namespace them.
That means namespacing Graphs needs to happen in the name of the Graph, whereas for any Asset created from this Graph the namespacing should happen via multi-component Asset keys, so having the namespacing also be part of the Asset name would be incorrect.

Example in pseudo code:
```
@graph()
def source1_users:
   ...

@graph()
def source2_users:
   ...

# Can't just call these graph's "users" since that would cause a collision
# because duplicate names are not allowed in a repository

# Current way AssetsDefinition.from_graph works
AssetsDefinition.from_graph(source1_users, key_prefix="source1")
# This would result in an Asset called "source1/source1_users") which is incorrect/doesn't make sense

# Expected way AssetsDefinition.from_graph should work
AssetsDefinition.from_graph(source1_users, key_prefix="source1", name="users")
# This would result in an Asset called "source1/users") as expected
```

This would also more closely align `AssetsDefinition.from_graph` with the `@asset` decorator because they both take (more of) the same arguments.

### Ideas of implementation

Maybe this somewhat older PR would solve this issue? https://github.com/dagster-io/dagster/pull/10152

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by OwenKephart** [comment]
Hi @simonvanderveldt! The current (somewhat inelegant) way of managing this at the moment is to use the `keys_by_output_name` parameter. The default output name is "result", so you can write:

```
AssetsDefinition.from_graph(
    source1_users,
    keys_by_output_name={"result": AssetKey(["source1", "users"])}
)
```

In general, the initial use for graph-backed assets was more focused around graph-backed assets which produced multiple assets, so the APIs get somewhat strained in terms of ergonomics when using them for just a single asset at a time, but this is something we can look into improving support for.

**Comment by simonvanderveldt** [comment]
@OwenKephart thanks for the pointer 👍  I'll try that out tomorrow and report back.

We're indeed using Graph backed Assets to orchestrate multiple steps to produce a single Asset. If it helps to know more about our usage/use-case just let us know, we can explain/share more either here/in a discussion or via Slack :)

**Comment by simonvanderveldt** [comment]
@OwenKephart Thanks a lot for the pointer, it's working now :+1: 
Got one small additional question just to check, the output name "result" is that coming from the default Out name as documented here https://docs.dagster.io/_apidocs/ops#dagster.Out?

> In general, the initial use for graph-backed assets was more focused around graph-backed assets which produced multiple assets, so the APIs get somewhat strained in terms of ergonomics when using them for just a single asset at a time, but this is something we can look into improving support for.

Regarding this part, I guess this is partially explained by `@asset` being singular/for a single Asset whereas `AssetsDefinition` can cover multiple assets?
Would it maybe make sense to have a way to create a single Asset from a graph that would be a closer match with `@asset`?

If something like this is being considered should I make a new issue for that and close this one? Or do you want to repurpose this one for that?

**Comment by sryza** [comment]
Hey @simonvanderveldt - I've got a PR in progress for a `@graph_asset` decorator: https://github.com/dagster-io/dagster/pull/10152. Would this help?

**Comment by simonvanderveldt** [comment]
> Hey @simonvanderveldt - I've got a PR in progress for a `@graph_asset` decorator: #10152. Would this help?

@sryza Yeah, that seems like it would be good fit for this indeed. Focused on producing a single Asset with a graph. The example code also looks nice because it focuses more on Assets instead of having to create a Graph to be able to generate an Asset from that. We don't really care about the Graph (instance) itself, it's just a means to create a single Asset from multiple steps.

I was wondering what the plan was with that PR? Do you still expect to progress in this direction?

**Comment by sryza** [comment]
Yup, got sidetracked with other things, but I can try to get it merged for the next release.

**Comment by simonvanderveldt** [comment]
This is fixed/addressed by #10152, thanks @sryza! :)
